viamonvif/discovery.go

```diff
@@ -1,241 +1,120 @@
-			return nil, fmt.Errorf("context canceled while connecting to ONVIF device: %s", xaddr)
-		}
-
-		logger.Debugf("Connecting to ONVIF device with URL: %s from interacted: %s", xaddr, iface.Name)
-
-		params := onvif.DeviceParams{
-			Xaddr: xaddr,
-		}
-		if username != "" {
-			params.Username = username
-		}
-		if password != "" {
-			params.Password = password
-		}
-
-		deviceInstance, err := onvif.NewDevice(params)
+			return zero, fmt.Errorf("context canceled while connecting to ONVIF device: %s", xaddr)
+		}
+		// This calls GetCapabilities
+		dev, err := device.NewDevice(device.DeviceParams{
+			Xaddr:    xaddr,
+			Username: cred.User,
+			Password: cred.Pass,
+		}, logger)
 		if err != nil {
-			logger.Warnf("failed to connect to ONVIF device: %v", err)
+			logger.Debugf("Failed to get camera capabilities from %s: %v", xaddr, err)
 			continue
 		}
 
-		cameraInfo, err := getCameraInfo(deviceInstance, username, password, logger)
+		cameraInfo, err := GetCameraInfo(dev, xaddr, cred, logger)
 		if err != nil {
 			logger.Warnf("Failed to get camera info from %s: %v", xaddr, err)
 			continue
 		}
-		cameraInfos = append(cameraInfos, cameraInfo)
+		// once we have addeed a camera info break
+		return cameraInfo, nil
 	}
-
-	return cameraInfos, nil
-}
-
-// if ifaceNames is nil or empty, we do all.
-func filterGoodInterface(all []net.Interface, ifaceNames []string, logger logging.Logger) []net.Interface {
-	var validInterfaces []net.Interface
-	for _, iface := range all {
-		if len(ifaceNames) > 0 && !slices.Contains(ifaceNames, iface.Name) {
-			logger.Debugf("skipping interface %s: not in list: %v", iface.Name, ifaceNames)
-			continue
-		}
-
-		if iface.Flags&net.FlagUp != 0 && iface.Flags&net.FlagMulticast != 0 && iface.Flags&net.FlagLoopback == 0 {
-			validInterfaces = append(validInterfaces, iface)
-		} else {
-			logger.Debugf("skipping interface %s: does not meet WS-Discovery requirements", iface.Name)
-		}
-	}
-	return validInterfaces
-}
-
-// DiscoverCameras performs WS-Discovery using the use-go/onvif discovery utility,
-// then uses ONVIF queries to get available RTSP addresses and supplementary info.
-// If ifaceNames is nil or empty, we try to discover over all available net interfaces on the machine.
-func DiscoverCameras(username, password string, logger logging.Logger, ifaceNames []string) (*CameraInfoList, error) {
-	var discoveredCameras []CameraInfo
-	ctx, cancelFunc := context.WithCancel(context.Background())
-	defer cancelFunc()
-
-	interfaces, err := net.Interfaces()
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve network interfaces: %w", err)
-	}
-
-	interfaces = filterGoodInterface(interfaces, ifaceNames, logger)
-
-	resultsChan := make(chan []CameraInfo, len(interfaces))
-	activeWorkers := sync.WaitGroup{}
-	for _, iface := range interfaces {
-		activeWorkers.Add(1)
-		utils.ManagedGo(func() {
-			cameraInfos, err := discoverCameraFromInterface(ctx, logger, iface, username, password)
```

Comment:
We don't need to run this in parallel

Reply:
turns out... we actually kinda do need to run this in parallel otherwise the mobile app hangs while trying to discover camers.

On the viam network the non parallel version takes 36 seconds to complete

The parallel version takes 13 seconds.

I think that is a big enough UX improvement to warrant the additional complexity.

I've added back parallelism which will:
1. Run WSDiscover on each interface in parallel (which will complete in 6 seconds regardless of the number of interfaces)
2. Query onvif endpoints on each device in parallel

Reply:
sean yu vindicated. okay with me.